### PR TITLE
DEV-5286 Add api factory

### PR DESCRIPTION
**Pull request by rizwansaeed**
# Pull Request Checklist

- [x] Read the [contributing guidelines](https://github.com/finbourne/lusid-sdk-python-preview/blob/master/CONTRIBUTING.md)
- [x] Tests pass
- [x] Raised the PR against the `develop` branch

# Description of the PR

Add an API factory to initialise API implementations with a token

### CODE REVIEW

File: sdk/lusid/utilities/api_client_factory.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                    else:
+                        #   switch to the '_with_http_info' implementation
+                        func = getattr(source_obj, f"{name}_with_http_info")
+                        result = func(*args, **kwargs)
+
+                    # pass the http info to caller
+                    callback(result[2])
+
+                    # return the dto
+                    return result[0]
```
[the remark this answers is not in the document]
Comment (MikeMcGarry) [reply sdk/lusid/utilities/api_client_factory.py]: Okay, it is the status code

File: sdk/lusid/utilities/api_client_factory.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            config.host = kwargs['api_url']
+
+            api_client = lusid.ApiClient(config,
+                                         header_name="X-LUSID-Application",
+                                         header_value=kwargs.get('app_name', None))
+
+        elif 'api_secrets_filename' in kwargs:
+            api_client = ApiClientBuilder.build(kwargs['api_secrets_filename'])
+        else:
+            raise Exception("Missing api initialisation values")
```
[the remark this answers is not in the document]
Comment (rizwansaeed) [reply sdk/lusid/utilities/api_client_factory.py]: done

File: sdk/lusid/utilities/api_client_factory.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        :return: Initalised LUSID API for the type passed in
+        """
+
+        def get_attribute_impl(source_obj, name):
+            """
+            Implementation of __getattribute__ that adds a decorator that adds a call_info
+            argument to return additional call stats
+            """
+
+            attr = super(obj, source_obj).__getattribute__(name)
```
[the remark this answers is not in the document]
Comment (rizwansaeed) [reply sdk/lusid/utilities/api_client_factory.py]: As discussed `obj` is the metaclass, have renamed the variable

File: sdk/tests/utilities/test_api_client_factory.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+from lusid.utilities import ApiClientFactory
+from utilities import TokenUtilities as tu, CredentialsSource
+
+
+class UnknownApi:
+    pass
+
+
+class UnknownImpl:
+    pass
```
[the remark this answers is not in the document]
Comment (rizwansaeed) [reply sdk/tests/utilities/test_api_client_factory.py]: The name 😏

File: sdk/tests/utilities/test_api_client_factory.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        factory = ApiClientFactory(api_secrets_filename=CredentialsSource.secrets_path())
+        with self.assertRaises(TypeError) as error:
+            factory.api(UnknownApi)
+        self.assertEqual(error.exception.args[0], "unknown api: UnknownApi")
+
+    def test_get_unknown_type_throws_exception(self):
+        factory = ApiClientFactory(api_secrets_filename=CredentialsSource.secrets_path())
+        with self.assertRaises(TypeError) as error:
+            factory.api(UnknownImpl)
+        self.assertEqual(error.exception.args[0], "unknown api: UnknownImpl")
```
[the remark this answers is not in the document]
Comment (rizwansaeed) [reply sdk/tests/utilities/test_api_client_factory.py]: There is a bit implied knowledge of the implementation in these tests. The first checks that any class which follows the naming convention of the API classes (xxxApi) but is not an API class is not passed in, the second that any class that is not an API class is not passed in

File: sdk/lusid/utilities/api_client_factory.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+            def wrapper(*args, **kwargs):
+
+                def is_http_info_method(m):
+                    return inspect.ismethod(m) and m.__name__.endswith('_with_http_info')
+
+                if kwargs.get('call_info', False) and True:
```
[the remark this answers is not in the document]
Comment (rizwansaeed) [reply sdk/lusid/utilities/api_client_factory.py]: done

File: sdk/lusid/utilities/api_client_factory.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                    callback = kwargs.pop('call_info')
```
[the remark this answers is not in the document]
Comment (rizwansaeed) [reply sdk/lusid/utilities/api_client_factory.py]: good shout 👍

File: sdk/tests/utilities/test_api_client_factory.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    pass
+
+
+class UnknownImpl:
+    pass
+
+
+class ApiFactory(unittest.TestCase):
+
+    def validate_api(self, api):
```
[the remark this answers is not in the document]
Comment (rizwansaeed) [reply sdk/tests/utilities/test_api_client_factory.py]: Good point. All accounts _should_ have the demo portfolios in so updated to check for > 0

**Pull request merged**